[CK_TILE] Add operator batched_transpose

fangche123 opened the pull request.

fangche123 committed:
add operator batched_transpose

poyenc commented on example/ck_tile/35_batched_transpose/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+add_executable(tile_example_batched_transpose EXCLUDE_FROM_ALL batched_transpose.cpp batched_transpose_api.cpp)
+target_include_directories(tile_example_batched_transpose PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/)
+target_compile_options(tile_example_batched_transpose PRIVATE -v --save-temps -Wno-gnu-line-marker)
```
Are those options essential? Or they are only used for debuging?

fangche123 replied:
Yes, these options are temporarily used for debug

poyenc committed:
Run remod.py under include/ck_tile & example/ck_tile folders

poyenc commented:
I just pushed re-formated files. Develops have to run remod.py script under **include/ck_tile** & **example/ck_tile** folders. 
The script will re-format source files and generate interface headers which include sub-directory headers.

poyenc commented on example/ck_tile/35_batched_transpose/batched_transpose_api.cpp:
```diff
@@ -0,0 +1,76 @@
+        {
+            BATCHED_TRANSPOSE_DISPATCH()
+        }
+        else if(t.layout == "NHWC")
```
Are those checks necessary? Either way you end-up calling the `BATCHED_TRANSPOSE_DISPATC()` macro

poyenc commented on example/ck_tile/35_batched_transpose/batched_transpose_api.hpp:
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2023, Advanced Micro Devices, Inc. All rights reserved.
```
update year.

fangche123 replied:
ok, I will fix it.

poyenc commented on example/ck_tile/35_batched_transpose/batched_transpose_api.cpp:
```diff
@@ -0,0 +1,76 @@
+#include "batched_transpose_api.hpp"
```
missed the copyright

poyenc commented on example/ck_tile/35_batched_transpose/CMakeLists.txt:
```diff
@@ -0,0 +1,3 @@
+add_executable(tile_example_batched_transpose EXCLUDE_FROM_ALL batched_transpose.cpp batched_transpose_api.cpp)
+target_include_directories(tile_example_batched_transpose PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/)
+target_compile_options(tile_example_batched_transpose PRIVATE -v --save-temps -Wno-gnu-line-marker)
```
you may need to define `TEST_BATCHED_TRANSPOSE_VERBOSE` macro here for testing.

poyenc commented on include/ck_tile/host/reference/reference_batched_transpose.hpp:
```diff
@@ -0,0 +1,58 @@
+                                              HostTensor<Type>& y,
+                                              std::string layout = "NCHW")
+{
+    // index_t rank = x.get_num_of_dimension();
```
Accidentally commented out?

poyenc commented on include/ck_tile/host/reference/reference_batched_transpose.hpp:
```diff
@@ -0,0 +1,58 @@
+namespace ck_tile {
+
+template <typename Type>
+CK_TILE_HOST void reference_batched_transpose(const HostTensor<Type>& x,
```
please document for the output layout.

poyenc commented on include/ck_tile/host/reference/reference_batched_transpose.hpp:
```diff
@@ -0,0 +1,58 @@
+
+template <typename Type>
+CK_TILE_HOST void reference_batched_transpose(const HostTensor<Type>& x,
+                                              HostTensor<Type>& y,
```
Please document for the output layout.

poyenc commented on include/ck_tile/ops/batched_transpose/kernel/batched_transpose_kernel.hpp:
```diff
@@ -0,0 +1,140 @@
+
+    CK_TILE_DEVICE void operator()(Kargs kargs) const
+    {
+        printf("in kernel. k.height:%d k.width:%d\n", kargs.height, kargs.width);
```
Forgot to remove debug code?

carlushuang committed:
[layernorm] hot fix (#1620)

* hot fix ln

* some rename

rocking5566 committed:
[Ck_tile] smoothquant (#1617)

* fix compile error

* fix typo of padding

* Add smoothquant op

* Add smoothquant instance library

* refine type

* add test script

* Re-generate smoothquant.hpp

* Always use 'current year' in copyright

* use Generic2dBlockShape instead

* Add vector = 8 instance back

* Find exe path automatically

* Simplify the api condition

* Remove debugging code

* update year

* Add blank line between function declaration

* explicitly cast return value to dim3

* refine return value

* Fix default warmup and repeat value

* Add comment

* refactor sommthquant cmake

* Add README

* Fix typo

---------

Co-authored-by: Po Yen, Chen <PoYen.Chen@amd.com>

illsilin committed:
Reduce build time. (#1621)

* disable fp8 gemm_universal on gfx90a and gfx908 by default

* fix cmake syntax

* fix clang format

* add ifdefs in amd_xdlops

* disable fp8 gemm instances on gfx90a by default

* update readme

carlushuang committed:
[CK_TILE] layernorm have more accurate residual (#1623)

* more accurate residual

* modify comment

* Fix literal case in README.md

---------

Co-authored-by: Po Yen Chen <PoYen.Chen@amd.com>

ThomasNing commented on example/ck_tile/35_batched_transpose/batched_transpose.cpp:
```diff
@@ -0,0 +1,310 @@
+auto create_args(int argc, char* argv[])
+{
+    ck_tile::ArgParser arg_parser;
+    arg_parser.insert("v", "1", "weather do CPU validation or not")
```
typo: whether

ThomasNing commented on example/ck_tile/35_batched_transpose/batched_transpose.cpp:
```diff
@@ -0,0 +1,310 @@
+#endif
+
+template <typename T>
+void dump_host_tensor_4d(const ck_tile::HostTensor<T>& x)
```
Please comment this out or put that to another file as a helper debug function. We could get a debug option in the argument input. WDYT?

ThomasNing commented on example/ck_tile/35_batched_transpose/batched_transpose.cpp:
```diff
@@ -0,0 +1,310 @@
+    int stride_dim3  = args.get_int("stride_dim3");
+    int seed         = args.get_int("seed");
+
+    std::string layout_in = "NCHW", layout_out = "NHWC";
```
Do we only support one kind of the memory layout from "NCHW" to "NHWC"? Theoretically, we could also support NWCH with another round of the transpose.

ThomasNing commented on example/ck_tile/35_batched_transpose/batched_transpose.cpp:
```diff
@@ -0,0 +1,310 @@
+    std::string layout_in = "NCHW", layout_out = "NHWC";
+    int stride_out_dim0 = stride_dim0, stride_out_dim1 = W * C, stride_out_dim2 = C,
+        stride_out_dim3 = 1;
+    if(stride_dim0 == C * H * W && stride_dim1 == 1 && stride_dim2 == W * C && stride_dim3 == C)
```
We should decide the stride configuration depending on the layout configuration right?

ThomasNing commented on example/ck_tile/35_batched_transpose/batched_transpose.hpp:
```diff
@@ -0,0 +1,130 @@
+
+#pragma once
+
+struct transpose_kernel_param_t
```
Comments on the pack_x pack_y & ediv_x ediv_y?

ThomasNing commented on example/ck_tile/35_batched_transpose/batched_transpose_api.cpp:
```diff
@@ -0,0 +1,76 @@
+                        batched_transpose_kargs a,
+                        ck_tile::stream_config s)
+{
+    if(t.type == "fp16")
```
You could put the type as a template in the transpose_calc function. PTAL gemm_kernel as a reference.

fangche123 committed:
update transpose: 1.add shared memory 2.support batch size > 1

fangche123 committed:
update transpose

bartekxk committed:
Temporary disable part of dynamic op conv instances (#1630)

* Temporary disable part of dynamic op conv instances

* fix

linsun12 committed:
Linsun/convint8 fwd instances (#1626)

Add instances for int8 grouped conv2d fwd
---------

Co-authored-by: root <root@dell300x-pla-t28-03.pla.dcgpu>
Co-authored-by: Bartłomiej Kocot <barkocot@amd.com>

jmmartinez committed:
[generate.py] Override blob list if it already exists (#1635)

Before, generate.py appended the list at the end of the output file.
When running the cmake configuration steps multiple times on the
examples, the blob list (such as fwd_blob_list.txt) would grow at every
configuration.
`library/src/tensor_operation_instance/gpu/mha/CMakeLists.txt` worked around
this issue by removing the output file if it exists.

Now, generate.py overrides the content of the output file.
There is no need for the workaround in the CMakeLists.txt;
and the issue is solved for the example projects too.

illsilin committed:
Make sure cmake can handle the xnack+/xnack- targets. (#1633)

* make sure cmake can handle xnack targets

* dont build xdl instances for gfx906:xnack-

* dont build xdl tests for gfx906:xnack-

darren-amd committed:
Statically Cast Pointer Offset (#1631)

* explicit cast ptr offset

* formating change

illsilin committed:
remove gfx940;gfx941 from default target lists (#1640)

andriy-ca committed:
Prevent instantiation of undefined FP8 operators. (#1639)

aledudek committed:
Generic threshold calculation after merge fixes (#1618)

* Generic threshold calculation add passing num of accums

* Generic threshold - after merge fixes

* Fix cmakelists

---------

Co-authored-by: Adam Osewski <19374865+aosewski@users.noreply.github.com>

rocking5566 committed:
Fix F16 type (#1583)

illsilin committed:
enable compilation for generic navi targets (#1645)

coderfeli committed:
[Ck tile] layernorm2d fwd optimize (#1637)

* optimze small N case using vec io and using rcp div

* [Ck_tile] layernorm, add param to control fastdiv; change generate codes and test pass

* [Ck_tile] fix blockSize compute in Generic2dBlockShape

* [Ck_tile]fix kfastfdiv template style

* [Ck_tile] layernorm, fix stype in review

---------

Co-authored-by: dummycoderfe <noplydummmycoder@163.com>

bartekxk committed:
Add generic instances for two stage conv bwd wei (#1643)

* Add generic instances for two stage conv bwd wei

* Update layout prefix

poyenc committed:
Fix 'sh' command compatibility of smoke_test_fwd.sh (#1553)

coderfeli committed:
Ck tile/moe sorting (#1624)

* add moe_sorting & check ok

* fix comments & typo

* Run remod.py under include/ck_tile & example/ck_tile directories

* format codes

* fix output ci check bug

* fix moe sorting readme and error commit file

* use magiv div to accelerate compute

* add an loop unroll for moe lds ops

* add extblocksnel to set zeros for moebufs

* [Ck_tile] moe set zero run ok, add size check and fix ref check

* [Ck_tile]fix moe_sorting fuse set_zero remod

* [Ck_tile] change name style, fix zero buffer size err, change folder

* [Ck_tile] moe_sorting: fix name style

* [Ck_tile] moe_sorting, remove useless params in traits

* [Ck_tile] change outputtile cnt * unit_size; change output buf alloc

---------

Co-authored-by: dummycoderfe <noplydummmycoder@163.com>
Co-authored-by: Po Yen, Chen <PoYen.Chen@amd.com>
Co-authored-by: carlushuang <carlus.huang@amd.com>

poyenc committed:
Return nullptr when block index is invalid (#1649)

valarLip committed:
[CK_TILE] add more stride for layernorm to support un-continuous Tensor (#1650)

* [CK_TILE] add more stride for layernorm to support un-continuous Tensor

* align CK coding style

* extend strides to layernrom expample

* clang-format...

illsilin committed:
restore collecting performance of mixed prec gemms (#1648)

ThomasNing committed:
[CK Tile] Improve the Layout, Padding, and Alignment features of CK Tile GEMM (#1651)

* Finished the feature

* Modified the test file

* Test case update

* addresss comment

* Addressed the review comment

* Fixed the CI error

fangche123 committed:
Handle the conflict

fangche123 committed:
Revert "Handle the conflict"

This reverts commit f9a58712def02214e5a43a6af5ee133e82d680fd, reversing
changes made to 2b6458ddf243904cecf4c54b48c9dafa60ff80df.
